translation of reference-glossary (#200)
reactjs/ru.reactjs.org: content/docs/reference-glossary.md

---
id: glossary
title: Словарь терминов React
layout: docs
category: Reference
permalink: docs/glossary.html

---

## Одностраничное приложение {#single-page-application}

> Одностраничное приложение — это приложение, которое состоит из единственной HTML страницы и прочих средств (таких как JavaScript и CSS), необходимых для успешной работы. Любое взаимодействие с главной или последующими ей страницами не требует контакта с сервером, что значит — страница не требует перезагрузки.

Reviewer: ```suggestion
Одностраничное приложение — это приложение, которое состоит из единственной HTML-страницы и прочих ресурсов (таких как JavaScript и CSS), необходимых для успешной работы. Любое взаимодействие с главной или последующими ей страницами не требует контакта с сервером, что значит — страница не требует перезагрузки.
```

P.S. assets - ресурсы

> Хоть React и позволяет создавать одностраничные приложения, это совсем не обязательно. React можно использовать для совершенствования небольших частей уже существующих сайтов, чтобы придать дополнительную интерактивность. Код, написанный в React, может мирно сосуществовать с разметкой отрендеренной на сервере с помощью PHP, или с любыми другими фронтенд-библиотеками. По сути, именно так React используется для Facebook.

Reviewer 2: ```suggestion
Хоть React и позволяет создавать одностраничные приложения, это совсем не обязательно. React можно использовать для совершенствования небольших частей уже существующих сайтов, чтобы придать дополнительную интерактивность. Код, написанный в React, может мирно сосуществовать с разметкой отрендеренной на сервере с помощью PHP, или с любыми другими фронтенд-библиотеками. По сути, именно так React и используется в Facebook.
```

> Хоть React и позволяет создавать одностраничные приложения, это совсем не обязательно. React можно использовать для совершенствования небольших частей уже существующих сайтов, чтобы придать дополнительную интерактивность. Код, написанный в React, может мирно сосуществовать с разметкой отрендеренной на сервере с помощью PHP, или с любыми другими фронтендовыми библиотеками. По сути, именно так React используется для Facebook.

Reviewer: ```suggestion
Хоть React и позволяет создавать одностраничные приложения, это совсем не обязательно. React можно использовать для совершенствования небольших частей уже существующих сайтов, чтобы придать дополнительную интерактивность. Код, написанный в React, может мирно сосуществовать с разметкой отрендеренной на сервере с помощью PHP, или с любыми другими фронтенд-библиотеками. По сути, именно так React используется для Facebook.
```

> ## ES6, ES2015, ES2016, и т.д. {#es6-es2015-es2016-etc}

Reviewer: ```suggestion
## ES6, ES2015, ES2016 и т.д. {#es6-es2015-es2016-etc}
```

> Все эти акронимы относяться к самым последним версиям спецификации стандарта ECMAScript — языка программирования лежащего в основе JavaScript. Версия ES6 (также известная как ES2015) включает много дополнений к предыдущим версиям: стрелочную функцию, классы, шаблонные строки, переменные `let` и `const`. Более подробно о конкретных версиях вы можно узнать [тут](https://ru.wikipedia.org/wiki/ECMAScript).

Reviewer: ```suggestion
Все эти акронимы относяться к самым последним версиям спецификации стандарта ECMAScript, реализацией которого является язык программирования JavaScript. Версия ES6 (также известная как ES2015) включает много дополнений к предыдущим версиям: стрелочные функции, классы, шаблонные строки, ключевые слова `let` и `const`. Более подробно о конкретных версиях вы можно узнать [тут](https://ru.wikipedia.org/wiki/ECMAScript).
```

Reviewer 3: относяться -> относятся

## Компиляторы {#compilers}

> Компайлер JavaScript преобразует и возвращает изменённый код. Самый известный случай использования — преобразование синтаксиса ES6 для поддержки в старых браузерах. В работе с React чаще всего используется [Babel](https://babeljs.io/)

Reviewer: ```suggestion
Компиляторы JavaScript преобразует и возвращает изменённый код. Самый известный случай использования — преобразование синтаксиса ES6 для поддержки в старых браузерах. В работе с React чаще всего используется [Babel](https://babeljs.io/)
```

> Компиляторы JavaScript преобразует и возвращает изменённый код. Самый известный случай использования — преобразование синтаксиса ES6 для поддержки в старых браузерах. В работе с React чаще всего используется [Babel](https://babeljs.io/)

Reviewer 2: ```suggestion
Компилятор JavaScript принимает на вход JavaScript код, преобразует его и возвращает в изменённом формате. Самый известный случай использования — преобразование синтаксиса ES6 для поддержки в старых браузерах. При работе с React чаще всего используется [Babel](https://babeljs.io/)
```

## Бандлеры {#bundlers}

> Бандлеры совмещают отдельные модули JavaScript и CSS для лучшей оптимизации в браузерах. В работе с React чаще всего используются [Webpack](https://webpack.js.org/) и [Browserify](http://browserify.org/).

Reviewer 2: ```suggestion
Бандлеры берут отдельные модули JavaScript и CSS и соединяют их в меньшее количество файлов, которые оптимзированы под браузеры. В работе с React чаще всего используются [Webpack](https://webpack.js.org/) и [Browserify](http://browserify.org/).
```

## Менеджер пакетов {#package-managers}

> Менеджер пакетов — это инструмент позволяющий управлять зависимостями в вашем проекте. [npm](https://www.npmjs.com/) и [Yarn](https://yarnpkg.com/) — менеджеры пакетов наиболее используемые в приложениях React. Оба являются клиентам реестра пакетов npm.

Reviewer 3: ```suggestion
Менеджер пакетов — это инструмент позволяющий управлять зависимостями в вашем проекте. [npm](https://www.npmjs.com/) и [Yarn](https://yarnpkg.com/) — менеджеры пакетов наиболее используемые в приложениях React. Оба являются клиентами реестра пакетов npm.
```

> Менеджер пакетов — это инструмент позволяющий управлять зависимостями в вашем проекте. [npm](https://www.npmjs.com/) и [Yarn](https://yarnpkg.com/) — менеджеры пакетов наиболее используемые в приложениях React. Оба являются клиентами реестра пакетов npm.

Reviewer 2: ```suggestion
Менеджер пакетов — это инструмент, позволяющий управлять зависимостями в вашем проекте. [npm](https://www.npmjs.com/) и [Yarn](https://yarnpkg.com/) — менеджеры пакетов наиболее используемые в приложениях React. Оба являются клиентами реестра пакетов npm.
```

## CDN {#cdn}

> Сеть доставки содержимого (англ. Content Delivery Network, CDN) — это сетевая инфраструктура, распространяющая кэшированный, статически контент через сеть серверов по всему миру.

Reviewer: ```suggestion
Сеть доставки содержимого (англ. Content Delivery Network, CDN) — это сетевая инфраструктура, распространяющая кешированный, статический контент через сеть серверов по всему миру.
```

## JSX {#jsx}

> JSX — расширение синтаксиса JavaScript, которое выглядит как шаблон, но имеет полную силу языка. В результате компиляции JSX и вызова `React.createElement()` возникают простые объекты — "элементы React". Чтобы ознакомиться с введением в JSX, [обратитесь к этой документации](/docs/introducing-jsx.html), более подробную информацию вы можете найти [здесь](/docs/jsx-in-depth.html).

Reviewer: ```suggestion
JSX — расширение синтаксиса JavaScript, которое выглядит как шаблон, но имеет полную силу языка. В результате компиляции JSX и вызова `React.createElement()` возникают простые объекты — «элементы React». Чтобы ознакомиться с введением в JSX, [обратитесь к соответствующему разделу документации](/docs/introducing-jsx.html), а более подробную информацию про JSX вы можете найти [здесь](/docs/jsx-in-depth.html).
```

> JSX — расширение синтаксиса JavaScript, которое выглядит как шаблон, но имеет полную силу языка. В результате компиляции JSX и вызова `React.createElement()` возникают простые объекты — «элементы React». Чтобы ознакомиться с введением в JSX, [обратитесь к соответствующему разделу документации](/docs/introducing-jsx.html), а более подробную информацию про JSX вы можете найти [здесь](/docs/jsx-in-depth.html).

Reviewer 4: Мне кажется, что в большинстве переводах мы старались переводить так "React-элемент"
```suggestion
JSX — расширение синтаксиса JavaScript, которое выглядит как шаблон, но имеет полную силу языка. В результате компиляции JSX и вызова `React.createElement()` возникают простые объекты — «React-элементы». Чтобы ознакомиться с введением в JSX, [обратитесь к соответствующему разделу документации](/docs/introducing-jsx.html), а более подробную информацию про JSX вы можете найти [здесь](/docs/jsx-in-depth.html).
```

> JSX — расширение синтаксиса JavaScript, которое выглядит как шаблон, но имеет полную силу языка. В результате компиляции JSX и вызова `React.createElement()` возникают простые объекты — «React-элементы». Чтобы ознакомиться с введением в JSX, [обратитесь к соответствующему разделу документации](/docs/introducing-jsx.html), а более подробную информацию про JSX вы можете найти [здесь](/docs/jsx-in-depth.html).

Reviewer 2: ```suggestion
JSX — расширение синтаксиса JavaScript. Этот синтаксис выглядит как язык шаблонов, но наделён всеми языковыми возможностями JavaScript. В результате компиляции JSX и вызова `React.createElement()` возникают простые объекты — «React-элементы». Чтобы ознакомиться с введением в JSX, [обратитесь к соответствующему разделу документации](/docs/introducing-jsx.html), а более подробную информацию про JSX вы можете найти [здесь](/docs/jsx-in-depth.html).
```

> React DOM использует camelCase стиль именования свойств вместо обычных имён HTML атрибутов. Например, в JSX `tabindex` становится `tabIndex`, а атрибут `class` становится `className` поскольку слово `class` уже зарезервировано в JavaScript:

Reviewer: ```suggestion
React DOM использует стиль именования camelCase для свойств вместо обычных имён HTML-атрибутов. Например, в JSX атрибут `tabindex` станет `tabIndex`. В то время как атрибут `class` записывается как `className`, поскольку слово `class` уже зарезервировано в JavaScript:
```

);
```  

## [Элементы](/docs/rendering-elements.html) {#elements}

> Элементы React — это составляющие блоки React-приложений. Их можно перепутать с более известной концепцией «компонентов», но в отличии от компонента, элемент описывает то, что вы хотите увидеть на экране. Элементы React иммутабельны.

Reviewer: ```suggestion
Элементы React — это составляющие блоки React-приложений. Их можно перепутать с более известной концепцией «компонентов», но в отличие от компонента, элемент описывает то, что вы хотите увидеть на экране. Элементы React иммутабельны.
```

> Элементы React — это составляющие блоки React-приложений. Их можно перепутать с более известной концепцией «компонентов», но в отличие от компонента, элемент описывает то, что вы хотите увидеть на экране. Элементы React иммутабельны.

Reviewer 4: ```suggestion
React-элементы — это составляющие блоки React-приложений. Их можно перепутать с более известной концепцией «компонентов», но в отличие от компонента, элемент описывает то, что вы хотите увидеть на экране. React-элементы иммутабельны.
```

```js
const element = <h1>Hello, world</h1>;
```

Обычно, элементы не используются напрямую, а возвращаются компонентами.

## [Компоненты](/docs/components-and-props.html) {#components}

> React компоненты — это маленькие, повторно используемые части кода, которые возвращают React елементы для отображения на странице.

Reviewer 3: ```suggestion
React компоненты — это маленькие, повторно используемые части кода, которые возвращают React-элементы для отображения на странице.
```

> React компоненты — это маленькие, повторно используемые части кода, которые возвращают React-элементы для отображения на странице.

Reviewer 4: ```suggestion
React-компоненты — это маленькие, повторно используемые части кода, которые возвращают React-элементы для отображения на странице.
```

> Самый простой компонент React — это простая JavaScript функция, которая возвращает елементы React:

Reviewer: ```suggestion
Самый простой компонент React — это простая функция JavaScript, которая возвращает элементы React:
```

> Самый простой компонент React — это простая функция JavaScript, которая возвращает элементы React:

Reviewer 4: ```suggestion
Самый простой React-компонент  — это простая функция JavaScript, которая возвращает элементы React:
```

> Самый простой React-компонент  — это простая функция JavaScript, которая возвращает элементы React:

Reviewer 2: ```suggestion
Самый простой React-компонент — это простая функция JavaScript, которая возвращает элементы React:
```

}
```

> Компоненты можно разбить на отдельные части в зависимости от выполняемой функции и использовать внутри других компонентов. Компоненты могут возвращать другие компоненты, массивы, строки и числа. Если какая-то часть интерфейса повторяется многократно (`Button`, `Panel`, `Avatar`) или сама по себе достаточно сложная (`App`, `FeedStory`, `Comment`), имеет смысл вынести её в независимый компонент. Имена компонентов всегда должны начинатся с заглавной буквы (`<Wrapper/>` **не** `<wrapper/>`). За более детальной информацией по рендерингу компонентов [обратитесь к этой документации](/docs/components-and-props.html#rendering-a-component).

Reviewer: ```suggestion
Компоненты можно разбить на отдельные части в зависимости от выполняемой функции и использовать внутри других компонентов. Компоненты могут возвращать другие компоненты, массивы, строки и числа. Если какая-то часть интерфейса повторяется многократно (`Button`, `Panel`, `Avatar`) или сама по себе достаточно сложная (`App`, `FeedStory`, `Comment`), имеет смысл вынести её в независимый компонент. Имена компонентов всегда должны начинатся с заглавной буквы (`<Wrapper/>`, а **не** `<wrapper/>`). За более детальной информацией по рендерингу компонентов [обратитесь к соответствующему разделу документации](/docs/components-and-props.html#rendering-a-component).
```

> Компоненты можно разбить на отдельные части в зависимости от выполняемой функции и использовать внутри других компонентов. Компоненты могут возвращать другие компоненты, массивы, строки и числа. Если какая-то часть интерфейса повторяется многократно (`Button`, `Panel`, `Avatar`) или сама по себе достаточно сложная (`App`, `FeedStory`, `Comment`), имеет смысл вынести её в независимый компонент. Имена компонентов всегда должны начинатся с заглавной буквы (`<Wrapper/>`, а **не** `<wrapper/>`). За более детальной информацией по рендерингу компонентов [обратитесь к соответствующему разделу документации](/docs/components-and-props.html#rendering-a-component).

Reviewer 2: ```suggestion
Компоненты можно разбить на отдельные части в зависимости от выполняемой функции и использовать внутри других компонентов. Компоненты могут возвращать другие компоненты, массивы, строки и числа. Если какая-то часть интерфейса повторяется многократно (`Button`, `Panel`, `Avatar`) или сама по себе достаточно сложная (`App`, `FeedStory`, `Comment`), имеет смысл вынести её в независимый компонент. Имена компонентов всегда должны начинатся с заглавной буквы (`<Wrapper/>`, а **не** `<wrapper/>`). За более детальной информацией о рендеринге компонентов [обратитесь к соответствующему разделу документации](/docs/components-and-props.html#rendering-a-component).
```

> ### [`пропсы`](/docs/components-and-props.html) {#props}

Reviewer: Если что-то заключено в \`\`, то в большинстве случае это не нужно переводить.

```suggestion
### [`props`](/docs/components-and-props.html) {#props}
```

> `props` (пропсы) — это входные данные компонентов React, передаваемые от компонента-родителя дочернему компоненту.

Reviewer 4: ```suggestion
`props` (пропсы) — это входные данные React-компонентов, передаваемые от родительского компонента дочернему компоненту.
```

> `пропсы` — это входные данные компонентов React, передаваемые от компонента-родителя дочернему компоненту.

Reviewer: ```suggestion
`props` (пропсы) — это входные данные компонентов React, передаваемые от компонента-родителя дочернему компоненту.
```

> Помните, `пропсы` предназначены только для чтения. Ни в каком случае их не следует менять:

Reviewer: ```suggestion
Помните, `props` предназначены только для чтения. Ни в каком случае их не следует изменять:
```

```js
// Неправильно!
props.number = 42;
```

> Если вам нужно поменять значение в ответ на пользовательский ввод или ответ сервера, используйте `состояние`.

Reviewer: ```suggestion
Если вам нужно поменять значение в ответ на пользовательский ввод или ответ сервера, используйте `state` (состояние).
```

### `props.children` {#propschildren}

> `props.children` доступны каждому компоненту. Это контент между открывающим и закрывающим тегом компонента. Например:

Reviewer 2: ```suggestion
В любом компоненте доступны `props.children`. Это контент между открывающим и закрывающим тегом компонента. Например:
```

```js
<Welcome>Hello world!</Welcome>
```

Строка `Hello world!` доступна в `props.children` в компоненте `Welcome`:

```js
function Welcome(props) {
  return <p>{props.children}</p>;
}
```

> Для компонентов-классов используйте `this.props.children`:

Reviewer: ```suggestion
Для классовых компонентов используйте `this.props.children`:
```

}
```

> ### [`состояние`](/docs/state-and-lifecycle.html#adding-local-state-to-a-class) {#state}

Reviewer: ```suggestion
### [`state`](/docs/state-and-lifecycle.html#adding-local-state-to-a-class) {#state}
```

> Компонент нуждается в `состоянии`, когда данные в нем со временем изменяются. Например, компоненту `Checkbox` может понадобится состояние `isChecked`, а компоненту `NewsFeed` необходимо отслеживать посты при помощи состояния `fetchedPosts`.

Reviewer: ```suggestion
Компонент нуждается в `state` (состоянии), когда данные в нем со временем изменяются. Например, компоненту `Checkbox` может понадобится состояние `isChecked`, а компоненту `NewsFeed` необходимо отслеживать посты при помощи состояния `fetchedPosts`.
```

> Самая большая разница между `состоянием` и `пропсами` состоит в том, что `пропсы` передаются от родителя потомку, а `состояние` управляется самим компонентом. Компонент не может воздействовать на `пропсы`, но может изменять `состояние`.  Для этого он должен вызвать `this.setState()`. Только классовые компоненты могут иметь состояние.

Reviewer: ```suggestion
Самая большая разница между `state` и `props` состоит в том, что `props` передаются от родителя потомку, а `state` управляется самим компонентом. Компонент не может воздействовать на `props`, но может изменять `state`.  Для этого он должен вызвать `this.setState()`. Только классовые компоненты могут иметь состояние.
```

> Для каждой отдельной части изменяемых данных должен существовать только один компонент, который "регулирует" изменение состояния. Не пытайтесь синхронизировать сосотояния двух разных компонентов. Вместо этого, [поднимите](/docs/lifting-state-up.html) их до ближайшего компонента-родителя и передайте через пропсы необходимым дочерним компонентам. 

Reviewer: ```suggestion
Для каждой отдельной части изменяемых данных должен существовать только один компонент, который «регулирует» изменение состояния. Не пытайтесь синхронизировать сосотояния двух разных компонентов. Вместо этого [поднимите](/docs/lifting-state-up.html) их до ближайшего компонента-родителя и передайте через пропсы необходимым дочерним компонентам. 
```

## [Методы жизненного цикла](/docs/state-and-lifecycle.html#adding-lifecycle-methods-to-a-class) {#lifecycle-methods}

Методы жизненного цикла — это настраиваемые функции, которые выполняются на различных этапах жизни компонента. Существуют специальные методы для первоначального рендеринга компонента в DOM ([монтирование](/docs/react-component.html#mounting)), его обновления, размонтирования и удаления.

>  ## [Управляемые](/docs/forms.html#controlled-components) и [неуправляемые компоненты](/docs/uncontrolled-components.html)

Reviewer: В соответствии со [словарём](https://github.com/reactjs/ru.reactjs.org/blob/master/TRANSLATION.md#%D1%81%D0%BE%D0%B3%D0%BB%D0%B0%D1%88%D0%B5%D0%BD%D0%B8%D0%B5-%D0%BF%D0%BE-%D0%BF%D0%B5%D1%80%D0%B5%D0%B2%D0%BE%D0%B4%D1%83-%D0%B3%D0%BB%D0%BE%D1%81%D1%81%D0%B0%D1%80%D0%B8%D0%B9):

```suggestion
 ## [Контролируемые](/docs/forms.html#controlled-components) и [неконтролируемые компоненты](/docs/uncontrolled-components.html)
```

В React существует два различных подхода для управления формами.

> Значением элемента формы input под контролем React — это *управляемый компонент*. Когда пользователь вводит данные в управляемый компонент, обработчик события изменений приходит в действие, и ваш код определяет допустим ли ввод (повторно рендерясь с обновленныем значением). Если повторный рендеринг не происходит, елемент формы остаётся без изменений.

Reviewer: Что-то не совсем ясно "Значением элемента формы (input) под контролем React...", может перефразировать иначе?

> *Неуправляемый компонент* работает как обычный  елемент формы вне React. Когда пользователь вводит данные в поле формы (поле ввода, выпадающий список и т.д.), обновлённая информация отображается без помощи React. Однако, это также значит, что некоторые значения не могут быть применены.

Reviewer: ```suggestion
*Неконтролируемый компонент* работает как обычный элемент формы вне React. Когда пользователь вводит данные в поле формы (поле ввода, выпадающий список и т.д.), обновлённая информация отображается без помощи React. Однако, это также значит, что некоторые значения не могут быть применены.
```

В большинстве случаев вам стоит использовать управляемые компоненты.

## [Ключи](/docs/lists-and-keys.html) {#keys}

«Ключ» – это специальный строковый атрибут, который нужно указывать при создании списка элементов. Ключи помогают React определять, какие элементы были изменены, добавлены или удалены. Их необходимо указывать, чтобы React мог сопоставлять элементы массива с течением времени.

> Ключи внутри массива должны быть уникальными только среди своих соседних элементов. Им не нужно быть уникальными во всем приложении или даже в общем компоненте.

Reviewer 3: ```suggestion
Ключи внутри массива должны быть уникальными только среди своих соседних элементов. Им не нужно быть уникальными во всём приложении или даже в общем компоненте.
```

> Не передавайте что-либо по типу `Math.random()` через ключи. Важно чтобы ключи имели "стабильное ID" при повторном рендеринге, чтобы React мог определить когда элементы добавлены, удалены или реорганизованы. В идеале, ключи должны соответствовать уникальному и стабильному идентификатору поступающему из ваших данных, такому как post.id`.

Reviewer: ```suggestion
Не передавайте что-либо по типу `Math.random()` через ключи. Необходимо создавать для ключей «постоянные идентификаторы» при повторных рендерингах, чтобы React мог определить когда элементы добавлены, удалены или пересортированы. В идеале, ключи должны соответствовать уникальному и постоянному идентификатору поступающему из ваших данных, такому как `post.id`.
```

## [Рефы](/docs/refs-and-the-dom.html) {#refs}

> React поддерживает особый атрибут, который можно добавить к любому компоненту. Атрибут `реф` может быть объектом, созданным при помощи [`React.createRef()` function](/docs/react-api.html#reactcreateref) или функцией-колбэком, или строкой (устаревшее API). Когда атрибут `реф` выступает в роли функции-колбэка, функция получает основной DOM элемент или экземпляр класса (в зависимости от типа элемента) в качестве аргумента. Это позволяет вам иметь прямой доступ к элементу DOM или экземпляру компонента.

Reviewer: ```suggestion
React поддерживает особый атрибут, который можно добавить к любому компоненту. Атрибут `ref` может быть объектом, созданным при помощи [функции `React.createRef()`](/docs/react-api.html#reactcreateref) или колбэком, либо же строкой (устаревшее API). Когда атрибут `ref` выступает в роли колбэка, функция получает основной DOM-элемент или экземпляр класса (в зависимости от типа элемента) в качестве аргумента. Это позволяет вам иметь прямой доступ к элементу DOM или экземпляру компонента.
```

Используйте рефы в исключительных случаях. Если вы часто обращаетесь к рефам за помощью, ознакомьтесь со [снисходящим потоком данных](/docs/lifting-state-up.html).

## [События](/docs/handling-events.html) {#events}

Обработка событий в React-элементах имеет некоторые синтактические особенности:

* Обработчики событий в React именуются в стиле camelCase вместо нижнего регистра.
* С JSX вы передаёте функцию как обработчик события вместо строки.

## [Согласование](/docs/reconciliation.html) {#reconciliation}

> Когда пропсы или состояние компонента изменяются, React сравнивает нововозвращенный и предыдуще отрендеренный элементы, и решает нужно ли обновлять DOM. Этот процесс называется "согласование".

Reviewer: ```suggestion
Когда пропсы или состояние компонента изменяются, React сравнивает нововозвращенный и предыдущий отрендеренные элементы, и решает нужно ли обновлять DOM. Этот процесс называется «согласование» (reconciliation).
```

### [`state`](/docs/state-and-lifecycle.html#adding-local-state-to-a-class) {#state}

> Компонент нуждается в `state` (состоянии), когда данные в нем со временем изменяются. Например, компоненту `Checkbox` может понадобится состояние `isChecked`, а компоненту `NewsFeed` необходимо отслеживать посты при помощи состояния `fetchedPosts`.

Reviewer 4: Я бы убрал уточнение в скобах, имхо, речь идёт именно о свойстве `state` у компонента.
```suggestion
Компонент нуждается в `state`, когда данные в нём со временем изменяются. Например, компоненту `Checkbox` может понадобится состояние `isChecked`, а компоненту `NewsFeed` необходимо отслеживать посты при помощи состояния `fetchedPosts`.
```

> Самая большая разница между `state` и `props` состоит в том, что `props` передаются от родителя потомку, а `state` управляется самим компонентом. Компонент не может воздействовать на `props`, но может изменять `state`.  Для этого он должен вызвать `this.setState()`. Только классовые компоненты могут иметь состояние.

Reviewer 4: @lex111 и @another-guy вроде бы было обсуждение, но не могу найти PR, речь шла о классовых компонентых и хуках, который позволяют добавить состояние в функциональные компоненты. Здесь встретилось похожее. В оригинале также есть это предложение. 

> Только классовые компоненты могут иметь состояние

Reviewer 2: @ntishkevich прости пожалуйста, я не понимаю, в чём заключается вопрос или совет. Можешь пояснить?

> Для каждой отдельной части изменяемых данных должен существовать только один компонент, который «регулирует» изменение состояния. Не пытайтесь синхронизировать сосотояния двух разных компонентов. Вместо этого [поднимите](/docs/lifting-state-up.html) их до ближайшего компонента-родителя и передайте через пропсы необходимым дочерним компонентам. 

Reviewer 2: ```suggestion
Для каждой отдельной части изменяемых данных должен существовать только один компонент, который «управляет» изменением состояния. Не пытайтесь синхронизировать состояния двух разных компонентов. Вместо этого [поднимите оба этих состояния](/docs/lifting-state-up.html) до ближайшего компонента-родителя и передайте через пропсы необходимым дочерним компонентам. 
```

## [Методы жизненного цикла](/docs/state-and-lifecycle.html#adding-lifecycle-methods-to-a-class) {#lifecycle-methods}

Методы жизненного цикла — это настраиваемые функции, которые выполняются на различных этапах жизни компонента. Существуют специальные методы для первоначального рендеринга компонента в DOM ([монтирование](/docs/react-component.html#mounting)), его обновления, размонтирования и удаления.

>  ## [Контролируемые](/docs/forms.html#controlled-components) и [неконтролируемые компоненты](/docs/uncontrolled-components.html)

Reviewer 4: Есть ишью #202  о переводе "контролируемые" -> "управляемые". Может получится сразу учесть это ишью и использовать уже нужный перевод.

```suggestion
 ## [Управляемые](/docs/forms.html#controlled-components) и [неуправляемые компоненты](/docs/uncontrolled-components.html)
```

В React существует два различных подхода для управления формами.

> Элемент формы input, управляемый React — это *контролируемый компонент*. Когда пользователь вводит данные в управляемый компонент, обработчик события изменений приходит в действие, и ваш код определяет допустим ли ввод (повторно рендерясь с обновленныем значением). Если повторный рендеринг не происходит, елемент формы остаётся без изменений.

Reviewer 4: Есть ишью #202 о переводе "контролируемые" -> "управляемые". Может получится сразу учесть это ишью и использовать уже нужный перевод.

```suggestion
Элемент формы input, управляемый React — это *управляемый компонент*. Когда пользователь вводит данные в управляемый компонент, обработчик события изменений приходит в действие, и ваш код определяет допустим ли ввод (повторно рендерясь с обновлённым значением). Если повторный рендеринг не происходит, елемент формы остаётся без изменений.
```

> Элемент формы input, управляемый React — это *контролируемый компонент*. Когда пользователь вводит данные в управляемый компонент, обработчик события изменений приходит в действие, и ваш код определяет допустим ли ввод (повторно рендерясь с обновленныем значением). Если повторный рендеринг не происходит, елемент формы остаётся без изменений.

Reviewer 2: ```suggestion
Элемент формы input, управляемый React — это *контролируемый компонент*. Когда пользователь вводит данные в управляемый компонент, обработчик события изменений приходит в действие, и ваш код определяет допустим ли ввод (повторно рендерясь с обновленныем значением). Если повторный рендеринг не происходит, элемент формы остаётся без изменений.
```

> *Неконтролируемый компонент* работает как обычный элемент формы вне React. Когда пользователь вводит данные в поле формы (поле ввода, выпадающий список и т.д.), обновлённая информация отображается без помощи React. Однако, это также значит, что некоторые значения не могут быть применены.

Reviewer 4: ```suggestion
*Неуправляемый компонент* работает как обычный элемент формы вне React. Когда пользователь вводит данные в поле формы (поле ввода, выпадающий список и т.д.), обновлённая информация отображается без помощи React. Однако, это также значит, что некоторые значения не могут быть применены.
```

> В большинстве случаев вам стоит использовать управляемые компоненты.

Reviewer 2: ```suggestion
В большинстве случаев вам следует использовать управляемые компоненты.
```

## [Ключи](/docs/lists-and-keys.html) {#keys}

«Ключ» – это специальный строковый атрибут, который нужно указывать при создании списка элементов. Ключи помогают React определять, какие элементы были изменены, добавлены или удалены. Их необходимо указывать, чтобы React мог сопоставлять элементы массива с течением времени.

Ключи внутри массива должны быть уникальными только среди своих соседних элементов. Им не нужно быть уникальными во всём приложении или даже в общем компоненте.

> Не используйте что-либо по типу `Math.random()` как ключи. Необходимо создавать для ключей «постоянные идентификаторы» при повторных рендерингах, чтобы React мог определить когда элементы добавлены, удалены или пересортированы. В идеале, ключи должны соответствовать уникальному и постоянному идентификатору поступающему из ваших данных, такому как `post.id`.

Reviewer 2: ```suggestion
Не используйте что-либо наподобие `Math.random()` как ключи. Необходимо создавать для ключей «стабильные идентификаторы» при повторных рендерингах, чтобы React мог определить когда элементы добавлены, удалены или переупорядочены. В идеале, ключи должны соответствовать уникальному и постоянному идентификатору поступающему из ваших данных, например `post.id`.
```

> Не передавайте что-либо по типу `Math.random()` через ключи. Необходимо создавать для ключей «постоянные идентификаторы» при повторных рендерингах, чтобы React мог определить когда элементы добавлены, удалены или пересортированы. В идеале, ключи должны соответствовать уникальному и постоянному идентификатору поступающему из ваших данных, такому как `post.id`.

Reviewer 4: ```suggestion
Не используйте что-либо по типу `Math.random()` как ключи. Необходимо создавать для ключей «постоянные идентификаторы» при повторных рендерингах, чтобы React мог определить когда элементы добавлены, удалены или пересортированы. В идеале, ключи должны соответствовать уникальному и постоянному идентификатору поступающему из ваших данных, такому как `post.id`.
```

## [Рефы](/docs/refs-and-the-dom.html) {#refs}

> React поддерживает особый атрибут, который можно добавить к любому компоненту. Атрибут `ref` может быть объектом, созданным при помощи [функции `React.createRef()`](/docs/react-api.html#reactcreateref) или колбэком, либо же строкой (устаревшее API). Когда атрибут `ref` выступает в роли колбэка, функция получает основной DOM-элемент или экземпляр класса (в зависимости от типа элемента) в качестве аргумента. Это позволяет вам иметь прямой доступ к элементу DOM или экземпляру компонента.

Reviewer 2: ```suggestion
React поддерживает особый атрибут, который можно прикрепить к любому компоненту. Атрибут `ref` может быть объектом, созданным при помощи [функции `React.createRef()`](/docs/react-api.html#reactcreateref) или колбэком, либо же строкой (устаревшее API). Когда в роли атрибута `ref` выступает колбэк, функция получает DOM-элемент, лежащий в основе компонента, или экземпляр класса (в зависимости от типа элемента) в качестве аргумента. Это позволяет вам иметь прямой доступ к элементу DOM или экземпляру компонента.
```

> Используйте рефы в исключительных случаях. Если вы часто обращаетесь к рефам за помощью, ознакомьтесь со [снисходящим потоком данных](/docs/lifting-state-up.html).

Reviewer 2: ```suggestion
Используйте рефы в исключительных случаях. Если вы часто обращаетесь к рефам за помощью, ознакомьтесь с [нисходящим потоком данных](/docs/lifting-state-up.html).
```

## [События](/docs/handling-events.html) {#events}

Обработка событий в React-элементах имеет некоторые синтактические особенности:

* Обработчики событий в React именуются в стиле camelCase вместо нижнего регистра.
* С JSX вы передаёте функцию как обработчик события вместо строки.

## [Согласование](/docs/reconciliation.html) {#reconciliation}

> Когда пропсы или состояние компонента изменяются, React сравнивает нововозвращенный и предыдущий отрендеренные элементы, и решает нужно ли обновлять DOM. Этот процесс называется «согласование» (reconciliation).

Reviewer 4: ```suggestion
Когда пропсы или состояние компонента изменяются, React сравнивает нововозвращённый и предыдущий отрендеренные элементы, и решает нужно ли обновлять DOM. Этот процесс называется «согласование» (reconciliation).
```
